icu4c/source/i18n/measfmt.cpp

```diff
@@ -770,71 +715,96 @@ static void appendRange(
-        appendRange(
-                smallestAmountFormatted,
-                0,
-                intFieldPosition.getBeginIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getBeginIndex(),
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                intFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-    } else {
-        appendTo.append(draft);
+    const NumberFormat *numberFormatter = numberFormat->get();
+    auto* numberFormatter2 = (NumberFormat*) numberFormatter->clone();
+    numberFormatter2->setMinimumIntegerDigits(2);
+
+    FormattedStringBuilder fsb;
+
+    UBool protect = FALSE;
+    const int32_t patternLength = pattern.length();
+    for (int32_t i = 0; i < patternLength; i++) {
+        char c = pattern[i];
+
+        // Also set the proper field in this switch
+        // We don't use DateFormat.Field because this is not a date / time, is a duration.
+        double value = 0;
+        switch (c) {
+            case 'H': value = hours; break;
+            case 'm': value = minutes; break;
+            case 's': value = seconds; break;
+        }
+
+        // There is not enough info to add Field(s) for the unit because all we have are plain
+        // text patterns. For example in "21:51" there is no text for something like "hour",
+        // while in something like "21h51" there is ("h"). But we can't really tell...
+        switch (c) {
+            case 'H':
+            case 'm':
+            case 's':
+                if (protect) {
+                    fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
+                } else {
+                    UnicodeString tmp;
+                    if ((i + 1 < patternLength) && pattern[i + 1] == c) { // doubled
+                        numberFormatter2->format(value, tmp, status);
+                        i++;
+                    } else {
+                        numberFormatter->format(value, tmp, status);
+                    }
+                    // TODO: Use proper Field
+                    fsb.append(tmp, (FormattedStringBuilder::Field) NULL, status);
+                }
+                break;
+            case '\'':
+                // '' is escaped apostrophe
+                if ((i + 1 < patternLength) && pattern[i + 1] == c) {
+                    fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
+                    i++;
+                } else {
+                    protect = !protect;
+                }
+                break;
+            default:
+                fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
+        }
     }
+
+    appendTo.append(fsb.toUnicodeString());
```

Comment:
You can do `toTempUnicodeString()` here.

Reply:
Done